apps/aehyperchains/src/aehc_parent_mng.erl

```diff
@@ -0,0 +1,333 @@
+    ok.
+
+%%%===================================================================
+%%%  State management
+%%%===================================================================
+
+get_track_by_connector(#state{tracks = Tracks}, Connector) ->
+    lists:keyfind(Connector, #track.connector, Tracks).
+
+set_track(#state{tracks = Tracks} = State, Track) ->
+    Connector = connector(Track),
+    Update = lists:keyreplace(Connector, #track.connector, Tracks, Track),
+    State#state{tracks = Update}.
+
+%%%===================================================================
+%%%  Connector's management
+%%%===================================================================
+
+-spec accept_connector(track()) -> boolean().
+accept_connector(Track) ->
+    try
+        [Criteria(Track) || Criteria <- fun accept_top_block/1, fun accept_block_by_hash/1, fun accept_send_tx/1],
+        ok
+    catch E:R ->
+        {error, {E, R}}
+    end.
+
+-spec start_connector(track()) -> {ok, pid()}.
+start_connector(Track) ->
+    aehc_connector_sup:start_child(connector(Track), args(Track), note(Track)).
+
+-spec terminate_connector(track()) -> ok.
+terminate_connector(Track) ->
+    aehc_connector_sup:terminate_child(connector(Track)).
+
+accept_top_block(Track) ->
+    %% Ability to request the current top block;
+    Con = connector(Track),
+    {ok, _} = aehc_connector:get_top_block(Con).
+
+accept_block_by_hash(Track) ->
+    %% Ability to request genesis hash track;
+    Con = connector(Track), Genesis = genesis_hash(Track),
+    {ok, _} = aehc_connector:get_block_by_hash(Con, Genesis).
+
+accept_send_tx(Track) ->
+    %% Ability to execute commitment call;
+    delegate_config() == undefined orelse
+        begin
+            Con = connector(Track),
+            Delegate = aec_keys:pubkey(),
+            KeyblockHash = aec_chain:top_key_block_hash(),
+            Payload = aehc_commitment_header:hash(aehc_commitment_header:new(Delegate, KeyblockHash)),
+            ok = aehc_connector:send_tx(Con, Payload)
+        end.
+
+%%%===================================================================
+%%%  Tracks management
+%%%===================================================================
+
+%% Initialization of view within Db (initial empty Db keep Genesis hash address as the top of parent view);
+-spec init_db_track(track()) -> track().
+init_db_track(Track) ->
+    GenesisHash = genesis_hash(Track),
+    Res = aehc_parent_db:get_parent_chain_track(GenesisHash),
+    TopBlockHash =
+        case Res of
+            undefined ->
+                aehc_parent_db:write_parent_chain_track(GenesisHash, GenesisHash),
+                GenesisHash;
+            _ when is_binary(Res) ->
+                Res
+        end,
+    Block = aehc_parent_db:get_parent_top_block(TopBlockHash),
+    init_track(Track, Block).
+
+%% Sync of track by fetching the current top;
+-spec sync_track(track()) -> track().
+sync_track(Track) ->
```

Comment:
This is blocking...

Reply:
Yep, you are right, but:

1) Are we eligible to proceed until one of the connectors isn't ready? Sounds simple, but actually I am not sure about;
2) I mention you transaction consequences below. We work around database and this table each time in a "freeze" state until call has executed. That means that a lot of parallelizing features don't impact here.

So, in general I agree that we have to decouple job into processes, and that's clear for me. But please also note the sync nature of some init procedures which we can't skip.

Reply:
Other parts of the node might want to inspect the status of the parent sync - especially a user wondering why the node isn't fully booting. Without entering the genserver loop quickly we essentially refuse to respond to requests about the status of the sync jobs. Please take a look at aec_sync which has a similar problem.

After sync is done the mining conductor should get notified that "Hey we are ready to participate in elections".